src/sentry/static/sentry/app/views/settings/account/accountSecurity/accountSecurityEnroll.jsx

```diff
@@ -210,152 +184,123 @@ class AccountSecurityEnroll extends AsyncView {
-            : t('Error sending SMS');
-          addErrorMessage(errorMessage);
-        }
+    try {
+      await this.api.requestPromise(this.enrollEndpoint, {data});
+    } catch (error) {
+      this._form = {};
+      const isSmsInterface = authenticator.id === 'sms';
+
+      this.setState({
+        hasSentCode: !isSmsInterface,
+      });
+
+      // Re-mount because we want to fetch a fresh secret
+      this.remountComponent();
+
+      addErrorMessage(
+        this.state.hasSentCode ? t('Incorrect OTP') : t('Error sending SMS')
       );
+
+      return;
+    }
+
+    if (!hasSentCode) {
+      // Just successfully finished sending OTP to user
+      this.setState({hasSentCode: true, loading: false});
+      addMessage(t('Sent code to %s', data.phone));
+    } else {
+      // OTP was accepted and SMS was added as a 2fa method
+      this.handleEnrollSuccess();
+    }
   };
 
   // Handle u2f device tap
-  handleU2fTap = data => {
-    return this.api
-      .requestPromise(`${ENDPOINT}${this.props.params.authId}/enroll/`, {
-        data: {
-          ...data,
-          ...this._form,
-          ...this.invite,
-        },
-      })
+  handleU2fTap = data =>
+    this.api
+      .requestPromise(this.enrollEndpoint, {data: {...data, ...this._form}})
       .then(this.handleEnrollSuccess, this.handleEnrollError);
-  };
 
   // Currently only TOTP uses this
-  handleSubmit = dataModel => {
-    const {authenticator} = this.state;
-
+  handleTotpSubmit = dataModel => {
     const data = {
       ...this._form,
       ...(dataModel || {}),
-      secret: authenticator.secret,
-      ...this.invite,
+      secret: this.state.authenticator.secret,
     };
 
-    this.setState({
-      loading: true,
-    });
+    this.setState({loading: true});
+
     this.api
-      .requestPromise(`${ENDPOINT}${this.props.params.authId}/enroll/`, {
-        method: 'POST',
-        data,
-      })
+      .requestPromise(this.enrollEndpoint, {method: 'POST', data})
       .then(this.handleEnrollSuccess, this.handleEnrollError);
   };
 
   // Handler when we successfully add a 2fa device
-  handleEnrollSuccess = () => {
-    this.loadOrganizationContext();
-    const authenticatorName =
-      (this.state.authenticator && this.state.authenticator.name) || 'Authenticator';
-    this.props.router.push('/settings/account/security');
-    openRecoveryOptions({
-      authenticatorName,
-    });
+  handleEnrollSuccess = async () => {
+    let next = '/settings/account/security';
+
+    // If we're pending approval of an invite, the user will have just joined
+    // the organization when completing 2fa enrollment. We should reload the
+    // organization context in that case to assign them to the org.
+    if (this.pendingInvitation) {
+      const org = await fetchOrganizationByMember(this.pendingInvitation.memberId, {
+        addOrg: true,
+        fetchOrgDetails: true,
+      });
+
+      next = `/organizations/${org.slug}/`;
```

Comment:
>After invite acceptance the accountSecurityEnroll view will now redirect to the organizations default page.

It was intentional to stay on the settings security page to encourage them to setup a backup phone number and/or save their recovery codes. It might be too easy to click out of this modal once you're on the issues page. What about loading the organization in context but not redirecting?

Reply:
There's a tricky bug with updating the organization index, which is why I ended up redirecting, to have the organization context properly update. It's actually broken in production right now.

I can dig deeper in, but I think it's not too bad to have the modal open while on the issues page

Reply:
I'm getting the modal closing before the issues page loads now. Are you seeing that on your side?

Reply:
Oh strange. Let me look closer, but I don't think I was seeing that.

Reply:
![image](https://user-images.githubusercontent.com/1421724/64567506-42c8c400-d30d-11e9-80dc-10bded5e06d7.png)

it actually stays open for me. Which one of the 2FAs do you setup?

Reply:
Ah, it's when you're redirected to the issues page. I'll rethink this.